---
title: 'Visualizations in R: box and violin plots'
author: "UC Davis Bioinformatics Core"
date: "`r Sys.Date()`"
output:
    html_document:
      keep_md: TRUE
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction to box and violin plots

Box plots and their cousins, violin plots, are a means of visualizing the distribution of a single variable. While a box plot is rectangular and marked with quantile lines, a violin plot has variable width designed to help the reader perceive the density of data. 

```{r example_gallery, echo=FALSE}

```

# Set-up

## Packages

As in the previous chapter, we will be working primarily within ggplot2. The box and violin geoms operate on the same "grammar" as the point geom. We'll move more quickly over concepts previously covered, and going into greater depth with new features as we come across them.

```{r packages}
library(dplyr)
library(tidyr)
library(magrittr)
library(kableExtra)
library(ggplot2)
library(ggrepel)
```

## Data

We'll use gene expression data from a single cell experiment for these plots. Typically this sort of data is stored in a complex structure which retains expression data and experiment metadata together in a single object. In this case, we will be using a Seurat object created in one of our single cell RNA-Seq workshops.

```{r data}
experiment.data <- readRDS("scRNA_workshop-05.rds")
```

This workshop does not aim to give a comprehensive understanding of the Seurat workflow and object; the relevant data for the purposes of box and violin plots are the scaled counts and experiment metadata.

Normalized expression data is stored in the "data" slot, and can be accessed as follows:

```{r scale.data, eval=FALSE}
experiment.data@assays$RNA$data
```

The metadata is stored in another slot within this object.

```{r metadata}
experiment.data@meta.data %>%
  slice(1:50) %>%
  kable() %>%
  kable_styling("striped", fixed_thead = TRUE) %>%
  scroll_box(height = "200px")
```

Let's create a smaller data frame containing the expression information for a select number of markers and a few key metadata values for each sample.

```{r markers}
markers <- c("SATB2", "NXPE1", "PDE3A", "CFTR", "HNF1A-AS1", "ADAMTSL1", "AC073050.1", "PID1", "NEO1", "XIST", "NR5A2", "AC019330.1", "CNTN4", "CNTN3", "SPON1", "LEFTY1")

markers <- markers[markers %in% rownames(experiment.data@assays$RNA$data)]

sc.data <- as.data.frame(t(as.matrix(experiment.data@assays$RNA$data[markers,])))
sc.data$cell <- rownames(sc.data)

metadata <- experiment.data@meta.data
metadata$cell <- rownames(metadata)
metadata <- select(metadata, cell, subcluster, subcluster_ScType_filtered)

sc.data <- inner_join(metadata, sc.data, by = "cell")

slice(sc.data, 1:50) %>%
  kable() %>%
  kable_styling("striped", fixed_thead = TRUE) %>%
  scroll_box(height = "200px")
rm(experiment.data, metadata, markers)
```


# Basic plot

The code below generates the most basic possible box plot; distribution of normalized expression values for a single gene across all cells sampled.

```{r basic_box}
ggplot(data = sc.data, mapping = aes(y = SATB2)) +
  geom_boxplot()
```

This is not as informative (or decorative) as it could be. Let's add information from the metadata.

## x axis

```{r subcluster}
ggplot(data = sc.data, mapping = aes(x = subcluster, y = SATB2)) +
  geom_boxplot()
```

Providing a categorical variable to the x axis produces a series of box plots corresponding to the levels of the variable.

## Color fill

You will often see filled box plots colored using the same variable as is used for the x axis. This can be useful across multi-panel visualizations to tie together the same samples visualized on different axes (e.g. a dimensionality reduction biplot and a box plot may share a color scheme to aid comprehension). However, it is possible to assign x and fill color independently.

```{r subcluster_type}
ggplot(data = sc.data, mapping = aes(x = subcluster, y = SATB2, fill = subcluster_ScType_filtered)) +
  geom_boxplot() +
  scale_fill_viridis_d(option = "turbo", name = "putative cell type") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

Coloring by putative cell type has highlighted the number of clusters for which cell type assignment was inconclusive. If we wish to emphasize a different contrast in our visualization, we can alter the factor leveling to change the color order. At the same time, we can set "Unknown" to NA and use the "na.value" argument to set the color to an unobtrusive light gray.

```{r}
sc.data$putative <- factor(gsub("Unknown", NA, sc.data$subcluster_ScType_filtered),
                           levels = c("Vascular endothelial cells",
                                      "Lymphoid cells",
                                      "Stromal cells",
                                      "Tuft cells",
                                      "Lymphatic endothelial cells",
                                      "Intestinal epithelial cells"))
ggplot(data = sc.data, mapping = aes(x = subcluster, y = SATB2, fill = putative)) +
  geom_boxplot() +
  scale_fill_viridis_d(option = "turbo",
                       name = "putative cell type",
                       begin = 0.1,
                       end = 0.9,
                       na.value = "gray90") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

If more than one marker is of interest, we can display these on a shared set of axes by reshaping the data frame to put the expression values in the same column. Associated gene symbols will be moved to another column, making the data frame much longer and narrower.

```{r pivot}
sc.pivot <- pivot_longer(sc.data, cols = SATB2:LEFTY1, names_to = "gene", values_to = "normalized.counts")
filter(sc.pivot, gene %in% c("PDE3A", "CFTR", "ADAMTSL1")) %>%
  ggplot(mapping = aes(x = subcluster, y = normalized.counts, fill = gene)) +
  geom_boxplot() +
  scale_fill_viridis_d() +
  theme_bw() +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(angle = 45))
```

Faceting is another approach to separating the data to show multiple markers.

```{r}
filter(sc.pivot, gene %in% c("PDE3A", "CFTR", "ADAMTSL1")) %>%
  ggplot(mapping = aes(x = subcluster, y = normalized.counts, fill = putative)) +
  geom_boxplot() +
  facet_wrap(~gene) +
  scale_fill_viridis_d(option = "turbo",
                       name = "putative cell type",
                       begin = 0.1,
                       end = 0.9,
                       na.value = "gray90") +
  theme_bw() +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(angle = 45))
```

All of this code works the same way on violin plots. Instead of quantile lines, the density of the distribution is indicated by the width of the plot.

```{r}
ggplot(data = sc.data, mapping = aes(y = SATB2, x = subcluster, fill = subcluster)) +
  geom_violin() +
  scale_fill_viridis_d(option = "plasma") + 
  guides(fill = "none") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45))
```

If needed, the axes can be flipped.

```{r}
ggplot(data = sc.data, mapping = aes(y = SATB2, x = subcluster, fill = subcluster)) +
  geom_violin() +
  scale_fill_viridis_d(option = "plasma") + 
  guides(fill = "none") +
  coord_flip() +
  theme_bw()
```

# Annotations

```{r}
library(ggsignif)

ggplot(data = sc.data, mapping = aes(x = putative, y = PDE3A, fill = putative)) +
  geom_boxplot() +
  scale_fill_viridis_d(option = "turbo",
                       begin = 0.1,
                       end = 0.9,
                       na.value = "gray90") +
   +
  guides(fill = "none") +
  theme_bw() +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```

# Composite plots

# Prepare for the next section

```{r prepare_next}
```

